src/qibo/quantum_info/utils.py

```diff
@@ -227,3 +228,68 @@ def hellinger_fidelity(prob_dist_p, prob_dist_q, validate: bool = False, backend
     distance = hellinger_distance(prob_dist_p, prob_dist_q, validate, backend=backend)
 
     return (1 - distance**2) ** 2
+
+
+def haar_integral(nqubits, t, samples, backend=None):
+    """Computes the integral over pure states over the Haar measure.
+
+    .. math::
+        \\int_{\\text{Haar}}\\left(|\\psi\\rangle\\right.\\left.\\langle\\psi|\\right)^{\\otimes t}d\\psi \\,
+
+    Args:
+        nqubits (int): Number of qubits.
+        t (int): index t to define the t-design.
+        samples (int): number of samples to estimate the integral.
+        backend (qibo.backends.abstract.Backend): Calculation engine.
+
+    Returns:
+        (array) : Estimation of the Haar integral.
+    """
+
+    from qibo.quantum_info import random_unitary
+
+    if backend == None:  # pragma: no cover
+        backend = GlobalBackend()
+
+    dim = 2**nqubits
+    randunit_density = backend.cast(np.zeros((dim**t, dim**t), dtype=complex))
+    for _ in range(samples):
+        haar_unit = random_unitary(dim, "haar", backend=backend)[:, 0].reshape(-1, 1)
+        rho = haar_unit @ np.conjugate(haar_unit.transpose())
+        randunit_density += ft.reduce(np.kron, [rho] * t)
+    return randunit_density / samples
+
+
+def pqc_integral(circuit, t, samples, backend=None):
+    """Computes the integral over pure states generated by uniformly sampling in the
+        parameter space described by the parameterized `circuit`.
+
+    .. math::
+        \\int_{\\Theta}\\left(|\\psi_{\\theta}\\rangle\\right.\\left.\\langle\\psi_{\\theta}|\\right)^{\\otimes t}d\\psi \\,
+
+    Args:
+        circuit (qibo.models.Circuit): Parametrized circuit.
+        t (int): index t to define the t-design.
+        samples (int): number of samples to estimate the integral.
+        backend (qibo.backends.abstract.Backend): Calculation engine.
+
+    Returns:
+        (array) : Estimation of the integral.
+    """
+
+    if backend == None:  # pragma: no cover
+        backend = GlobalBackend()
+
+    circuit.density_matrix = True
+    dim = 2**circuit.nqubits
+    randunit_density = backend.cast(np.zeros((dim**t, dim**t), dtype=complex))
+    nparams = 0
+    for gate in circuit.queue:
+        if hasattr(gate, "trainable") and gate.trainable:
+            nparams += len(gate.parameters)
+    for _ in range(samples):
+        params = np.random.uniform(-np.pi, np.pi, nparams)
+        circuit.set_parameters(params)
+        rho = backend.execute_circuit(circuit).state()
+        randunit_density += ft.reduce(np.kron, [rho] * t)
+    return randunit_density / samples
```

Comment:
it's good for readability to have a blank like before the `return` (in all functions)